go/integration_test/processor/processor_gas_test.go

```diff
@@ -0,0 +1,503 @@
+// Copyright (c) 2024 Fantom Foundation
+//
+// Use of this software is governed by the Business Source License included
+// in the LICENSE file and at fantom.foundation/bsl11.
+//
+// Change Date: 2028-4-16
+//
+// On the date above, in accordance with the Business Source License, use of
+// this software will be governed by the GNU Lesser General Public License v3.
+
+package processor
+
+import (
+	"fmt"
+	"strings"
+	"testing"
+
+	"github.com/Fantom-foundation/Tosca/go/processor/floria"
+	"github.com/Fantom-foundation/Tosca/go/tosca"
+	op "github.com/ethereum/go-ethereum/core/vm"
+)
+
+func getGasTestScenarios() map[string]Scenario {
+	const (
+		codePrice = 3 + 3
+		excessGas = 100
+	)
+
+	allTestCases := map[string]Scenario{
+		"ValueTransferExact": {
+			Before: WorldState{
+				{1}: Account{Balance: tosca.NewValue(100), Nonce: 4},
+			},
+			Transaction: tosca.Transaction{
+				Sender:    tosca.Address{1},
+				Recipient: &tosca.Address{2},
+				GasLimit:  floria.TxGas,
+				Value:     tosca.NewValue(3),
+				Nonce:     4,
+			},
+			After: WorldState{
+				{1}: Account{Balance: tosca.NewValue(97), Nonce: 5},
+				{2}: Account{Balance: tosca.NewValue(3)},
+			},
+			Receipt: tosca.Receipt{
+				Success: true,
+				GasUsed: floria.TxGas,
+			},
+		},
+		"ValueTransferSufficient": {
+			Before: WorldState{
+				{1}: Account{Balance: tosca.NewValue(100), Nonce: 4},
+			},
+			Transaction: tosca.Transaction{
+				Sender:    tosca.Address{1},
+				Recipient: &tosca.Address{2},
+				GasLimit:  floria.TxGas + excessGas,
+				Value:     tosca.NewValue(3),
+				Nonce:     4,
+			},
+			After: WorldState{
+				{1}: Account{Balance: tosca.NewValue(97), Nonce: 5},
+				{2}: Account{Balance: tosca.NewValue(3)},
+			},
+			Receipt: tosca.Receipt{
+				Success: true,
+				GasUsed: floria.TxGas + excessGas/10, // 1/10th of the gas is always consumed
+			},
+		},
+		"ValueTransferInsufficient": {
+			Before: WorldState{
+				{1}: Account{Balance: tosca.NewValue(100), Nonce: 4},
+			},
+			Transaction: tosca.Transaction{
+				Sender:    tosca.Address{1},
+				Recipient: &tosca.Address{2},
+				GasLimit:  floria.TxGas - 1,
+				Value:     tosca.NewValue(3),
+				Nonce:     4,
+			},
+			After: WorldState{
+				{1}: Account{Balance: tosca.NewValue(100), Nonce: 4},
+			},
+			Receipt: tosca.Receipt{
+				Success: false,
+				GasUsed: 0,
```

Comment:
We once started discussing this but did not get to a conclusion; I would assume that in such a case the GasLimit might have to be returned, since this corresponds to the actual gas cost, right? After all, I would assume that the gas bought by the user at the begin of the transaction is not returned.

This would also make this case handled identically to a case where one is running out of gas during a contract execution.

Reply:
I also thought this is very surprising in the current reference implementation, but I thought it has to be correct because the whole history passed. I changed the reference implementation on this branch, it still has been confirmed to be compatible with the history.

Reply:
It might very likely be the case that in case of a failure the `GasUsed` field is ignored in Aida and the full gas consumption is just hard-coded.

It might also be that `receipts` are not checked at all in `aida-vm`.